### ISSUE TITLE
Firefox compatibility

### ISSUE THREAD

**Issue by fregante** [pixiebrix/pixiebrix-extension#490]
I'll post here whatever issues I find in Firefox, so they can be tracked in one place for the time being.

---

It appears that `panels.elements` is not supported and our code does attempt to use it, which may be why the dev tools are not loading in Firefox currently:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Extending_the_developer_tools#devtools.panels

---

`externally_connectable` is not available: 
- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/externally_connectable 
- https://bugzilla.mozilla.org/show_bug.cgi?id=1319168

---

Other issues found in the label: https://github.com/pixiebrix/pixiebrix-extension/labels/firefox

**Comment by twschiller** [comment]
Thanks for starting this issue!

I think this was an attempt at re-creating some of the `externally_connectable` behavior using the content script and `document.defaultView.postMessage`: https://github.com/pixiebrix/pixiebrix-extension/blob/ead1833564fff81288a195503b2dea57d3b55d2e/src/contentScript/externalProtocol.ts

Haven't tested it to see if it still works

**Comment by fregante** [comment]
I found out why it's not working in Firefox: it sends messages when the content script isn't yet ready.

In Chrome, `runtime.sendMessage` is available immediately because it communicates directly to the background, but Firefox can't communicate until both the content script and script.js are injected.

Here you can see pb.com sending 5 requests before the content script is even executed. 

<img width="610" alt="Screen Shot 6" src="https://user-images.githubusercontent.com/1402241/123408780-9afeb200-d5d7-11eb-92e7-b80e75e03f51.png">

Unfortunately `document_start` only fixes the issue some of the time, this needs to be fixed in pb.com so that it waits before sending those messages.

I'm not entirely sure of how it's implemented because pb.com didn't receive the improvements in https://github.com/pixiebrix/pixiebrix-extension/pull/420 and the source code is basically obfuscated by the [regenerator](https://github.com/facebook/regenerator).

I suppose we have 2 options:

- message queue that is read by script.js (e.g. a global array)
- have a global callback function that is called by script.js

Alternatives?

**Comment by fregante** [comment]
`CONNECT_PAGE` doesn't seem to be registered among the page script handlers. Related:

- https://github.com/pixiebrix/pixiebrix-extension/issues/578

It seems that it's because `messaging/external.ts` is not included in `script.js` and if I try to include I get a rollbar error which causes the whole file to fail, it seems.

```
 Uncaught Error: Permission denied to access property "_rollbarOldOnError"
    captureUncaughtExceptions moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:63981
    setupUnhandledCapture moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:63981
    m moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:63981
    init moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:63981
    ts moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:67396
    __webpack_require__ moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:72586
    ts moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:64173
    __webpack_require__ moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:72586
    ts moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:66258
    __webpack_require__ moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:72586
    <anonymous> moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:72680
    <anonymous> moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:72862
    <anonymous> moz-extension://fb57a464-0fcd-bd4d-8c0c-7da43ba6341a/script.js:72864
```

**Comment by fregante** [comment]
Firefox is missing:

### `onMessage.hasListeners()` (it only has `hasListener(fn)`)

No bugs open yet: https://bugzilla.mozilla.org/buglist.cgi?quicksearch=haslisteners

https://github.com/pixiebrix/pixiebrix-extension/blob/c34b29d67f09523d80c7ba1b68a5ebaa06c9293c/src/background/devtools/external.ts#L92

### `browser.permissions` is not available in devtoolsPanel.html

Somewhat related to https://bugzilla.mozilla.org/show_bug.cgi?id=1493396 (sidebar, not fixed) and https://github.com/sindresorhus/refined-github/issues/840#issuecomment-349936195 (preferences page, fixed)

Possible solutions:

- Direct them to the options page to add a permission (a button on the current page [isn't permitted](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/User_actions), unless we include an iframe pointing to an extension page)
- [webext-domain-permission-toggle](https://github.com/fregante/webext-domain-permission-toggle) for Firefox users + copy [this guide](https://fregante.github.io/webext-domain-permission-toggle/?name=PixieBrix&icon=https%3A%2F%2Fraw.githubusercontent.com%2Fpixiebrix%2Fpixiebrix-extension%2Fmain%2Fimg%2Flogo-small.svg) into `<PermissionPane>`, without having to go through the options page.


---

More issues coming, probably around script.js messaging.

**Comment by twschiller** [comment]
@fregante I'm finding it hard to follow the blockers in this issue so that we can find workaround/prioritize them. I think it would help to track these as separate issues with a "Firefox" tag as we'll need a separate discussion for each (e.g., tracking the Firefox bug tracker, and discussing work arounds)

> Firefox is missing: onMessage.hasListeners()

What is this impacting? 

> browser.permissions

I think the webext-domain-permission-toggle approach makes sense for Firefox issues. We'll just need to style the guide to match the existing style within the devtools

**Comment by fregante** [comment]
Blocked by:

- [ ] https://github.com/pixiebrix/pixiebrix-app/issues/202

Next:

- ensuring that the `externally_connectable` polyfill works
- https://github.com/pixiebrix/pixiebrix-extension/labels/firefox

**Comment by fregante** [comment]
Blocked by:

- [ ] https://github.com/pixiebrix/pixiebrix-app/issues/302

**Comment by twschiller** [comment]
Local-only mode probably won't happen for awhile (although it would be very useful to help support Selenium testing). Are we blocked by anything else on Firefox? It looks like https://github.com/pixiebrix/pixiebrix-app/issues/302 is the last piece?

**Comment by fregante** [comment]
Yeah local-only is no longer a blocker, only https://github.com/pixiebrix/pixiebrix-app/issues/302 is, for now. I've been testing the extension gradually so I still haven't been able to test anything past the saving due to the blocker.

**Comment by fregante** [comment]
I think the Firefox version is in a decent state now 🎉  and I can't find more issues other than what I have already opened: https://github.com/issues?q=is%3Aopen+user%3Apixiebrix+label%3Afirefox+

Please test it in Firefox and open specific issues if any.
